bokehjs/src/coffee/models/widgets/data_table.coffee

```diff
@@ -66,7 +67,10 @@ export class DataProvider
       [column.sortCol.field, if column.sortAsc then 1 else -1]
 
     if cols.length == 0
-      cols = [[DTINDEX_NAME, 1]]
+      if @model.index_column
+        cols = [[DTINDEX_NAME, 1]]
+      else
+        return 0
```

Comment:
It should be simply `return`, aka. skip the rest of the code. Though I'm not sure if this change is needed at all, because `cols` is empty, we should be restoring the index order. It doesn't matter if index column is visible or not.